Functions/introduction_to_functions_and_scope.md

```diff
@@ -0,0 +1,262 @@
+```
+
+## Other types of functions
+
+1. **Higher-Order Functions**:
+   Functions that take other functions as arguments or return functions as results. This allows to eEnable powerful patterns like function composition and currying.
+
+2. **Factory Functions**:
+   A factory function is a function that returns an object. Used for creating multiple instances of similar objects with different properties. Encapsulates object creation logic within a function, promoting code reusability.
+
+3. **Pure Functions**:
+   Pure functions are functions that return the same output for the same input and have no side effects. They depend solely on their input parameters, without modifying any external state or variables. Pure functions are predictable, making it easier to reason about their behavior and test them reliably. They contribute to code maintainability, immutability, and enable functional programming paradigms.
+
+4. **Callbacks**
+   Functions that are passed to other functions as arguments are often called callbacks.
+
+## Scope
+
+If you are writing a Javascript program that runs in a browser and you declare a variable outside of any function, you are declaring a variable in the **global scope**. That variable can be accessed anywhere. This is in contrast to variables that are declared within functions, which are said to be in local scope.
+
+The fact that global variables can be accessed anywhere means that they can easily be overwritten accidentally, resulting in chaos. For this and other reasons it is best to minimize the number of global variables you create. There is rarely any reason to create more than one. There is a global object and it is this object that this refers to in global scope. Variables that are declared in the global scope with **var** keyword automatically become properties of the global object.
+
+### Undeclared assignment
+
+In both global and local scope, if you assign a value to a variable without declaring that variable, you create a global variable/property.
+
+```bash
+a = 100;
+a; //100;
+this.a //100
+```
+
+This is easy to do accidentally. To prevent such mistakes, undeclared assignment throws an error in strict mode.
+
+```bash
+'use strict';
+a = 100; //ERROR
+```
+
+### Hoisting
+
+In both global and local scope, declarations of variables with var are processed prior to the execution of all other code, including assignments. This is called hoisting. Hoisting makes the following three code snippets equivalent.
+
+```bash
+var a = 100;
+var b = 200;
+```
+
+```bash
+var a, b;
+a = 100;
+b = 200;
+```
+
+```bash
+a = 100;
+b = 200;
+var a, b;
+```
+
+Note that in the third example above, there is no error in strict mode because there is no undeclared assignment.
+
+### Local scope
+
+Variables declared in a function are in local scope. They are not accessible outside of the function.
+
+```js
+var fn = function () {
+    var n = 0;
+    console.log(typeof n);
+};
+
+fn(); //logs 'number'
+console.log(typeof n);
```

Comment:
I guess it is missing the: //logs undefined

Reply:
Exactly, missed that while copying from dillinger.io